File: Android/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/uiperformance/UIPerformanceManager.java
package com.didichuxing.doraemonkit.kit.uiperformance;

import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.support.v4.app.Fragment;
import android.view.View;
import android.view.ViewGroup;

import com.blankj.utilcode.util.ActivityUtils;
import com.didichuxing.doraemonkit.model.ViewInfo;
import com.didichuxing.doraemonkit.util.LifecycleListenerUtil;
import com.didichuxing.doraemonkit.util.LogHelper;
import com.didichuxing.doraemonkit.util.UIUtils;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by wanglikun on 2019-06-27
 */
public class UIPerformanceManager implements LifecycleListenerUtil.LifecycleListener {
    private static final String TAG = "UIPerformanceManager";
    private Canvas mPerformanceCanvas;
    private List<PerformanceDataListener> mListeners = new ArrayList<>();

    private static class Holder {
        private static UIPerformanceManager INSTANCE = new UIPerformanceManager();
    }

    public static UIPerformanceManager getInstance() {
        return Holder.INSTANCE;
    }

    private UIPerformanceManager() {

    }

    public void start(Context context) {
        Bitmap canvasBitmap = Bitmap.createBitmap(UIUtils.getWidthPixels(), UIUtils.getHeightPixels(), Bitmap.Config.ARGB_8888);
        mPerformanceCanvas = new Canvas(canvasBitmap);
        LifecycleListenerUtil.registerListener(this);
    }

    public void stop() {
        mListeners.clear();
        mPerformanceCanvas = null;
        LifecycleListenerUtil.unRegisterListener(this);
    }

    public List<ViewInfo> getViewInfos(Activity activity) {
        if (activity == null) {
            LogHelper.d(TAG, "resume activity is null");
            return new ArrayList<>();
        }
        if (activity.getWindow() == null) {
            LogHelper.d(TAG, "resume activity window is null");
            return new ArrayList<>();
        }

        return getViewInfos(UIUtils.getDokitAppContentView(activity));
    }

    private List<ViewInfo> getViewInfos(View view) {
        List<ViewInfo> infos = new ArrayList<>();
        traverseViews(view, infos, 0);
        return infos;
    }

    private void traverseViews(View view, List<ViewInfo> infos, int layerNum) {
        if (view == null) {
            return;
        }
        layerNum++;
        if (view instanceof ViewGroup) {
            int childCount = ((ViewGroup) view).getChildCount();
            if (childCount != 0) {
                for (int index = childCount - 1; index >= 0; index--) {
                    traverseViews(((ViewGroup) view).getChildAt(index), infos, layerNum);
                }
            }
        } else {
            long startTime = System.nanoTime();
            view.draw(mPerformanceCanvas);
            long endTime = System.nanoTime();
            float time = (endTime - startTime) / 10_000 / 100f;
            //LogHelper.d(TAG, "drawTime: " + time + " ms");
            ViewInfo viewInfo = new ViewInfo(view);
            viewInfo.drawTime = time;
            viewInfo.layerNum = layerNum;
            infos.add(viewInfo);
        }
    }

    public void addListener(PerformanceDataListener listener) {
        mListeners.add(listener);
    }

    public void removeListener(PerformanceDataListener listener) {
        mListeners.remove(listener);
    }

    @Override
    public void onActivityResumed(Activity activity) {
        for (PerformanceDataListener listener : mListeners) {
            listener.onRefresh(getViewInfos(activity));
        }
    }

    @Override
    public void onActivityPaused(Activity activity) {

    }

    @Override
    public void onFragmentAttached(Fragment f) {
        for (PerformanceDataListener listener : mListeners) {
            listener.onRefresh(getViewInfos(f.getActivity()));
        }
    }

    @Override
    public void onFragmentDetached(Fragment f) {
        for (PerformanceDataListener listener : mListeners) {
            listener.onRefresh(getViewInfos(f.getActivity()));
        }
    }

    public interface PerformanceDataListener {
        void onRefresh(List<ViewInfo> viewInfos);
    }

    /**
     * 初始化时直接显示显示层级
     */
    public void initRefresh() {
        for (PerformanceDataListener listener : mListeners) {
            listener.onRefresh(getViewInfos(ActivityUtils.getTopActivity()));
        }
    }

}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/uiperformance/UIPerformanceManager.kt
package com.didichuxing.doraemonkit.kit.uiperformance

import android.app.Activity
import android.graphics.Bitmap
import android.graphics.Canvas
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.blankj.utilcode.util.ActivityUtils
import com.didichuxing.doraemonkit.model.ViewInfo
import com.didichuxing.doraemonkit.util.LifecycleListenerUtil
import com.didichuxing.doraemonkit.util.LifecycleListenerUtil.registerListener
import com.didichuxing.doraemonkit.util.LifecycleListenerUtil.unRegisterListener
import com.didichuxing.doraemonkit.util.LogHelper.d
import com.didichuxing.doraemonkit.util.UIUtils.getDokitAppContentView
import com.didichuxing.doraemonkit.util.UIUtils.heightPixels
import com.didichuxing.doraemonkit.util.UIUtils.widthPixels
import java.util.*

/**
 *
 * Desc:ui层级管理类
 * <p>
 * Date: 2020-06-15
 * Company:
 * Updater:
 * Update Time:
 * Update Comments:
 *
 * Author: pengyushan
 */
class UIPerformanceManager private constructor() : LifecycleListenerUtil.LifecycleListener {
    private var mPerformanceCanvas: Canvas? = null
    private val mListeners: MutableList<PerformanceDataListener> = ArrayList()

    private object Holder {
        val INSTANCE = UIPerformanceManager()
    }

    /**
     *
     * Desc:开始跟踪显示界面信息
     * <p>
     * Author: pengyushan
     * Date: 2020-06-15
     */
    fun startTrack() {
        val canvasBitmap = Bitmap.createBitmap(widthPixels, heightPixels, Bitmap.Config.ARGB_8888)
        mPerformanceCanvas = Canvas(canvasBitmap)
        registerListener(this)
    }

    /**
     *
     * Desc:停止跟踪显示界面信息
     * <p>
     * Author: pengyushan
     * Date: 2020-06-15
     */
    fun stopTrack() {
        mListeners.clear()
        mPerformanceCanvas = null
        unRegisterListener(this)
    }

    fun addListener(listener: PerformanceDataListener) {
        mListeners.add(listener)
    }

    fun removeListener(listener: PerformanceDataListener) {
        mListeners.remove(listener)
    }

    override fun onActivityResumed(activity: Activity?) {
        for (listener in mListeners) {
            listener.onRefresh(UIPerformanceUtil.getActivityViewInfo(activity,mPerformanceCanvas))
        }
    }

    override fun onActivityPaused(activity: Activity?) {}
    override fun onFragmentAttached(f: Fragment?) {
        for (listener in mListeners) {
            listener.onRefresh(UIPerformanceUtil.getActivityViewInfo(f!!.activity,mPerformanceCanvas))
        }
    }

    override fun onFragmentDetached(f: Fragment?) {
        for (listener in mListeners) {
            listener.onRefresh(UIPerformanceUtil.getActivityViewInfo(f!!.activity,mPerformanceCanvas))
        }
    }

    interface PerformanceDataListener {
        fun onRefresh(viewInfos: List<ViewInfo>?)
    }

    /**
     * 初始化时直接显示显示层级
     */
    fun initRefresh() {
        for (listener in mListeners) {
            listener.onRefresh(UIPerformanceUtil.getActivityViewInfo(ActivityUtils.getTopActivity(),mPerformanceCanvas))
        }
    }

    companion object {
        private const val TAG = "UIPerformanceManager"
        @JvmStatic
        val instance = Holder.INSTANCE
    }
}